[ML] Refactor AutodetectMemoryLimitIT to use memory limit constants and improve model size assertions

valeriy42 opened the pull request:
Previously, the upper bound for model memory checks was set in absolute terms, which is not easy to understand and is brittle. I adjusted the assertions to ensure that the memory usage does not exceed 5% of the memory limit. Additionally, on Linux, we now report the process size (see #131981), which includes approximately 20 MB of native code overhead. I made handling this overhead more explicit. 

More details:
- Removed muted tests for `testManyDistinctOverFields` and `testTooManyByAndOverFields`.
- Introduced constants for memory limits in `AutodetectMemoryLimitIT`.
- Updated assertions to check effective model size against calculated limits.

Closes #132308
Closes #132310
Closes #132611

valeriy42 committed:
Refactor AutodetectMemoryLimitIT to use memory limit constants and improve model size assertions

- Removed muted tests for `testManyDistinctOverFields` and `testTooManyByAndOverFields`.
- Introduced constants for memory limits in `AutodetectMemoryLimitIT`.
- Updated assertions to check effective model size against calculated limits.

Commit:
[CI] Auto commit changes from spotless

valeriy42 committed:
fix compilation errors

valeriy42 committed:
Merge branch 'tests/AutodetectMemoryLimitIT' of https://github.com/valeriy42/elasticsearch into tests/AutodetectMemoryLimitIT

jan-elastic commented on x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/AutodetectMemoryLimitIT.java:
```diff
@@ -78,15 +83,22 @@ public void testTooManyPartitions() throws Exception {
+        assertThat(
+            getEffectiveModelSize(modelSizeStats.getModelBytes()),
+            allOf(
+                greaterThan(ByteSizeValue.ofMb(memoryLimit).getBytes() * 0.15),
```
Why such a low 15%?

The model is supposed to hit either the soft or hard limit, so I imagine the minimum should be a lot more.

jan-elastic reviewed (approved):
Generally look good. Some small comments.

valeriy42 committed:
Refactor AutodetectMemoryLimitIT to use memory limit variables

- Changed variable names from `memoryLimit` to `memoryLimitMb` for clarity.
- Updated memory limit assertions to reflect the new variable naming.
- Ensured consistency in memory limit usage across multiple test cases.

valeriy42 committed:
Merge branch 'main' into tests/AutodetectMemoryLimitIT

valeriy42 committed:
Merge branch 'main' into tests/AutodetectMemoryLimitIT

valeriy42 committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into tests/AutodetectMemoryLimitIT

valeriy42 commented on x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/AutodetectMemoryLimitIT.java:
```diff
@@ -225,12 +229,12 @@ public void testManyDistinctOverFields() throws Exception {
-        assertThat(modelSizeStats.getModelBytes(), lessThan(120500000L));
-        assertThat(modelSizeStats.getModelBytes(), greaterThan(70000000L));
+        assertThat(getEffectiveModelSize(modelSizeStats.getModelBytes()), lessThan(ByteSizeValue.ofMb(memoryLimitMb).getBytes() * 1.05));
         assertThat(modelSizeStats.getMemoryStatus(), equalTo(ModelSizeStats.MemoryStatus.HARD_LIMIT));
```
The reporting of memory usage is different on different plattforms: on Linux and Windows we report actual process memory usage, while on MacOS this information is not available and hence we report estimated memory usage. This makes the test for a specific memory limit brittle. To improve robustness, I only check that the job is in the `hard_limit` state.

valeriy42 commented on x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/AutodetectMemoryLimitIT.java:
```diff
@@ -78,15 +81,16 @@ public void testTooManyPartitions() throws Exception {
-        assertThat(modelSizeStats.getModelBytes(), greaterThan(24000000L));
+
+        assertThat(getEffectiveModelSize(modelSizeStats.getModelBytes()), lessThan(ByteSizeValue.ofMb(memoryLimitMb).getBytes() * 1.05));
         assertThat(
```
Only check for `hard_limit` to increase test robustness.

valeriy42 committed:
Remove unnecessary overhead calculations and assertions in AutodetectMemoryLimitIT tests

valeriy42 committed:
Merge branch 'tests/AutodetectMemoryLimitIT' of https://github.com/valeriy42/elasticsearch into tests/AutodetectMemoryLimitIT

valeriy42 committed:
Merge branch 'main' into tests/AutodetectMemoryLimitIT
